lexos/models/kmeans_model.py

```diff
@@ -0,0 +1,455 @@
+        k_means_index = k_means.fit_predict(reduced_data)
+
+        # This is important, such that plot and table results are consistent.
+        sorted_k_means_unique_index = sorted(set(k_means_index))
+
+        # Get file names.
+        labels = np.array([self._id_temp_label_map[file_id]
+                           for file_id in self._doc_term_matrix.index.values])
+
+        # Get a list of lists of file names based on the cluster result.
+        cluster_labels = [labels[np.where(k_means_index == index)]
+                          for index in sorted_k_means_unique_index]
+
+        # Get a list of lists of file coordinates based on the cluster result.
+        cluster_values = [reduced_data[np.where(k_means_index == index)]
+                          for index in sorted_k_means_unique_index]
+
+        # Get a list of centroid results based on the cluster result.
+        centroid_values = [np.mean(cluster, axis=0, dtype="float_")
+                           for cluster in cluster_values]
+
+        # Find the decision boundary of the graph.
+        # First find list of x, y coordinates.
+        x_value, y_value = reduced_data[:, 0], reduced_data[:, 1]
+
+        # Find min, max for x and then calculate bounds and step.
+        x_min, x_max = x_value.min(), x_value.max()
+        x_low_bound = x_min - (x_max - x_min) / 5
+        x_up_bound = x_max + (x_max - x_min) / 5
+        # Increase 200 will make lines smoother.
+        x_step = (x_up_bound - x_low_bound) / 200
+
+        # Find min, max for y and then calculate bounds and step.
+        y_min, y_max = y_value.min(), y_value.max()
+        y_low_bound = y_min - (y_max - y_min) / 5
+        y_up_bound = y_max + (y_max - y_min) / 5
+        # Increase 200 will make lines smoother.
+        y_step = (y_up_bound - y_low_bound) / 200
+
+        # Find x, y mesh grids, decrease the step to make lines smoother.
+        x_mesh_grid, y_mesh_grid = \
+            np.meshgrid(np.arange(x_low_bound, x_up_bound, x_step),
+                        np.arange(y_low_bound, y_up_bound, y_step))
+
+        # Find K Means predicted z values.
+        z_value = k_means.predict(np.c_[x_mesh_grid.ravel(),
+                                        y_mesh_grid.ravel()])
+
+        # Reshape Z value based on shape of x mesh grid.
+        z_value = z_value.reshape(x_mesh_grid.shape)
+
+        # Draw the regions with heat map.
+        # This method could be updated once plotly better support polygons.
+        voronoi_regions = [go.Heatmap(x=x_mesh_grid[0][:len(z_value)],
+                                      y=x_mesh_grid[0][:len(z_value)],
+                                      z=z_value,
+                                      hoverinfo="skip",
+                                      showscale=False,
+                                      colorscale='YlGnBu')]
+
+        # Pick a color for following scatter plots.
+        color = cl.scales["10"]["qual"]["Paired"]
```

Comment:
This function is awfully long (L165-L292, way over the 30-line limit for functions). Is there a way to split this function into several functions?